File: lib/data/disk.cc
#include "lib/core/globals.h"
#include "lib/data/disk.h"
#include "lib/data/image.h"
#include "lib/data/layout.h"
#include "lib/data/sector.h"
#include "protocol.h"

namespace
{
    struct pair_to_range_t
    {
        template <typename I>
        friend constexpr auto operator|(
            std::pair<I, I> const& pr, pair_to_range_t)
        {
            return std::ranges::subrange(pr.first, pr.second);
        }
    };
    inline constexpr pair_to_range_t pair_to_range{};
}

Disk::Disk(): image(std::make_shared<Image>()) {}

Disk::Disk(
    const std::shared_ptr<const Image>& image, const DiskLayout& diskLayout):
    image(image)
{
    std::multimap<CylinderHead, std::shared_ptr<const Sector>>
        sectorsGroupedByTrack;
    for (const auto& sector : *image)
        sectorsGroupedByTrack.insert(
            std::make_pair(sector->physicalLocation.value(), sector));

    const auto sectorLocations = std::views::keys(sectorsGroupedByTrack);
    for (const auto& physicalLocation :
        std::set(sectorLocations.begin(), sectorLocations.end()))
    {
        const auto& ptl =
            diskLayout.layoutByPhysicalLocation.at(physicalLocation);
        const auto& ltl = ptl->logicalTrackLayout;

        auto decodedTrack = std::make_shared<Track>();
        decodedTrack->ltl = ltl;
        decodedTrack->ptl = ptl;
        tracksByPhysicalLocation.insert(
            std::make_pair(physicalLocation, decodedTrack));

        for (auto& [ch, sector] :
            sectorsGroupedByTrack.equal_range(physicalLocation) | pair_to_range)
        {
            decodedTrack->allSectors.push_back(sector);
            decodedTrack->normalisedSectors.push_back(sector);
            sectorsByPhysicalLocation.insert(
                std::make_pair(physicalLocation, sector));
        }
    }
}

File: java/com/cowlark/fluxengine/data/Disk.java
package com.cowlark.fluxengine.data;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ListMultimap;
import java.util.Set;
import java.util.TreeSet;

/**
 * A disk, being the result of reading a physical disk, ported from
 * lib/data/disk.h and lib/data/disk.cc.
 */
public class Disk
{
    public final ListMultimap<CylinderHead, Track> tracksByPhysicalLocation =
            ArrayListMultimap.create();
    public final ListMultimap<CylinderHead, Sector> sectorsByPhysicalLocation =
            ArrayListMultimap.create();
    public Image image = null;

    /* 0 if the period is unknown (e.g. if this Disk was made from an image). */
    public double rotationalPeriod = 0;

    public Disk()
    {
        image = new Image();
    }

    public Disk(Image image, DiskLayout diskLayout)
    {
        this.image = image;

        ListMultimap<CylinderHead, Sector> sectorsGroupedByTrack = ArrayListMultimap.create();
        for (Sector sector : image)
            sectorsGroupedByTrack.put(sector.physicalLocation, sector);

        Set<CylinderHead> sectorLocations = new TreeSet<>();
        for (CylinderHead ch : sectorsGroupedByTrack.keySet())
            sectorLocations.add(ch);

        for (CylinderHead physicalLocation : sectorLocations)
        {
            PhysicalTrackLayout ptl = diskLayout.layoutByPhysicalLocation.get(physicalLocation);
            LogicalTrackLayout ltl = ptl.logicalTrackLayout;

            Track decodedTrack = new Track();
            decodedTrack.ltl = ltl;
            decodedTrack.ptl = ptl;
            tracksByPhysicalLocation.put(physicalLocation, decodedTrack);

            for (Sector sector : sectorsGroupedByTrack.get(physicalLocation))
            {
                decodedTrack.allSectors.add(sector);
                decodedTrack.normalisedSectors.add(sector);
                sectorsByPhysicalLocation.put(physicalLocation, sector);
            }
        }
    }
}
